import-automation/progress-dashboard-rest/app/resource/import_attempt.py

```diff
@@ -0,0 +1,149 @@
+NOT_FOUND_ERROR = 'attempt_id not found'
+
+
+class ImportAttempt(flask_restful.Resource):
+    """Base import attempt resource.
+
+    Attributes:
+        database: A database service for storing import attempts
+    """
+    parser = reqparse.RequestParser()
+    optional_fields = (
+        ('attempt_id',), ('branch_name',), ('repo_name',), ('pr_number', int),
+        ('import_name',), ('provenance_url',), ('provenance_description',),
+        ('email',), ('status',), ('time_created',), ('logs', dict, 'append')
+    )
+    utils.add_optional_fields(parser, optional_fields)
+
+    def __init__(self):
+        self.database = import_attempt_database.ImportAttemptDatabase()
+
+
+class ImportAttemptByID(ImportAttempt):
+    """API for managing import attempts by attempt_id associated with
+    '/import/<string:attempt_id>'."""
+
+    def put(self, attempt_id):
+        """Overwrites or creates an attempt with the given ID.
+
+        time_created must be in ISO 8601 format with timezone UTC+0, e.g.
+        '2020-06-30T04:28:53.717569+00:00'. If time_created is not specified,
+        the time at which the request is processed will be used.
+
+        status must be one of the allowed status defined by ImportStatus. If
+        status is not specified, 'created' will be used.
+
+        Args:
+            attempt_id: ID string of the attempt
+
+        Returns:
+            The created import attempt if successful as a dict. Otherwise,
+            (error message, error code), where the error message is a string
+            and the error code is an int.
+        """
+        args = self.parser.parse_args()
+        if attempt_id != args.setdefault('attempt_id', attempt_id):
+            return ID_NOT_MATCH_ERROR, http.HTTPStatus.CONFLICT
+        status = args.setdefault('status', ImportStatus.CREATED.value)
+        if status not in IMPORT_STATUS:
+            return ('Import status {} is not allowed'.format(status),
+                    http.HTTPStatus.FORBIDDEN)
+
+        import_attempt = self.database.get_by_id(attempt_id, make_new=True)
+        args.setdefault('logs', [])
+        args.setdefault('time_created', utils.utctime())
+        import_attempt.update(args)
+
+        return self.database.save(import_attempt)
+
+    def get(self, attempt_id):
+        """Retrieves an attempt by its attempt_id.
+
+        Args:
+            attempt_id: ID of the attempt as a string
+
+        Returns:
+            The import attempt with the ID if successful as a dict. Otherwise,
+            (error message, error code), where the error message is a string
+            and the error code is an int.
+        """
+        import_attempt = self.database.get_by_id(attempt_id)
+        if not import_attempt:
+            return NOT_FOUND_ERROR, http.HTTPStatus.NOT_FOUND
+        return import_attempt
+
+    def patch(self, attempt_id):
+        """Modifies the value of a field of an existing attempt.
+
+        attempt_id is forbidden to modify.
```

Comment:
Perhaps change the argument specification to "as a constant string", omitting this comment.